File: java_code/src/java/com/artvb/iq/problem1_1/UniqChars.java
package com.artvb.iq.problem1_1;

import java.util.Arrays;

/**
 * Created by artburkart on 10/19/14.
 */
public class UniqChars {
    /**
     * Sorts String to check for redundancy of chars in given String.
     *
     * O(nlogn) Sort array with dual-pivot quicksort.
     *
     * @param str String to check for redundancy of chars
     * @return true if all chars unique, otherwise false
     */
    private static boolean Method1(String str) {
        // Input cannot be null, has no unique chars if length is 0,
        // and must be unique if length 1
        UniqChars.assertNotNull(str);
        if (str.length() == 0) { return false; }
        if (str.length() == 1) { return true; }

        // Derive char array from input String and sort
        char[] chars = str.toCharArray();
        Arrays.sort(chars);

        for (int i = 0; i < chars.length - 1; i++) {
            if (chars[i] == chars[i+1]) { return false; }
        }
        return true;
    }

    /**
     * Uses boolean mask to check for redundancy of chars in given String.
     *
     * O(n) - Iterate through chars twice to find duplicate.
     *
     * @param str String to check for redundancy of chars
     * @return true if all chars unique, otherwise false
     */
    private static boolean Method2(String str) {
        // Input cannot be null, has no unique chars if length is 0,
        // and must be unique if length 1
        UniqChars.assertNotNull(str);
        if (str.length() == 0) { return false; }
        if (str.length() == 1) { return true; }

        // Find maximum char ordinal value for mask initialization
        int max = Integer.MIN_VALUE;
        char[] chars = str.toCharArray();
        for (char c : chars) {
            max = (c > max) ? c : max;
        }

        // Initialize mask with length of (max + 1)
        boolean[] mask = new boolean[max + 1];
        for (char c : chars) {
            if (mask[c]) { return false; }
            mask[c] = true;
        }
        return true;
    }

    /**
     * Uses bit twiddling to check for redundancy in given String
     *
     * Note: chars in String must fall within an ordinal range of 32
     *
     * @param str String to check for redundancy of chars
     * @return true if all chars unique, otherwise false
     */
    private static boolean Method3(String str) {
        // Input cannot be null, has no unique chars if length is 0,
        // and must be unique if length 1
        UniqChars.assertNotNull(str);
        if (str.length() == 0) { return false; }
        if (str.length() == 1) { return true; }

        int min = Integer.MAX_VALUE;
        int max = Integer.MIN_VALUE;

        // Find and min and max ordinal char values in String
        char[] chars = str.toCharArray();
        for (char c: chars) {
            min = (c < min) ? c : min;
            max = (c > max) ? c : max;
        }

        // If range of ordinal char values is greater than 32, then this
        // method will not work on the string
        if (max - min > Integer.SIZE - 1) {
            String err = String.format("Range between ordinal char values " +
                "is larger than maximum (%d).", Integer.SIZE - 1);
            throw new IllegalArgumentException(err);
        }

        int mask = 0;
        int marker, offset;
        for (char c: chars) {
            offset = c - min;
            marker = 1 << offset;
            if ((marker & mask) > 0) { return false; }
            mask |= marker;
        }
        return true;
    }

    /**
     * Asserts that an object not be null.
     *
     * @param o Object to check whether null
     */
    private static void assertNotNull(Object o) {
        if (o == null) {
            throw new NullPointerException("Given input must not be null");
        }
    }


    /**
     * Selects algorithm for determining redundancy in given String.
     *
     * @param str String to check if contains only unique chars
     * @param method determines which algorithm to use
     */
    public static boolean uniqChars(String str, int method) {
        UniqChars.assertNotNull(method);
        // Use different algorithms based on method option provided
        switch (method) {
            case UniqChars.SORTED:
                return UniqChars.Method1(str);
            case UniqChars.BOOL:
                return UniqChars.Method2(str);
            case UniqChars.BIT:
                return UniqChars.Method3(str);
            default:
                throw new IllegalArgumentException("Given `method` is not valid.");
        }
    }

    // Available methods for checking uniqueness of a String
    public static final int SORTED = 1;
    public static final int BOOL = 2;
    public static final int BIT = 3;
}

File: python_code/src/Chapter1/Problem1/UniqChars.py
class UniqChars:
    """Provides different algorithms for checking no redundancy
    exists in a given String
    """
    def __init__(self):
        """Initializes the UniqChars class
        """
        pass

    def __assert_str_or_unicode(self, obj, expected):
        """Asserts an object not be None.

        Keyword arguments:
        obj -- the object to check for nullness
        expected -- the expected type of object
        """
        if not isinstance(obj, str) or isinstance(obj, unicode):
            raise AttributeError("""\
                Given input, {}, must be an instance of str or unicode
                """.format(expected))

    def __method1(self, string):
        """Sorts string, then
        returns True if redundant chars present, otherwise False

        O(nlogn) sorts array with dual-pivot quicksort

        Keyword arguments:
        string -- str to check for redundancy of chars
        """
        self.__assert_str_or_unicode(string, str)
        if (len(string) == 0):
            return False
        if (len(string) == 1):
            return True

        sorted_string = sorted(string)
        for i, ch in enumerate(sorted_string[:-1]):
            if ch == sorted_string[i + 1]:
                return False
        return True

    def __method2(self, string):
        """Creates boolean mask, then
        returns True if redundant chars present, otherwise False

        O(n) iterates through chars twice

        Keyword arguments:
        string -- str to check for redundancy of chars
        """
        self.__assert_str_or_unicode(string, str)
        if (len(string) == 0):
            return False
        if (len(string) == 1):
            return True

        max_val = ord(max(string))
        boolean = [False for i in range(max_val + 1)]
        for c in string:
            if boolean[ord(c)]:
                return False
            boolean[ord(c)] = True
        return True

    def __method3(self, string):
        """Creates binary mask, then
        returns True if redundant chars present, otherwise False

        O(n) iterates through chars once

        Keyword arguments:
        string -- str to check for redundancy of chars
        """
        self.__assert_str_or_unicode(string, str)
        if (len(string) == 0):
            return False
        if (len(string) == 1):
            return True

        mask = 0
        for c in string:
            marker = 1 << ord(c)
            if (marker & mask > 0):
                return False
            mask |= marker
        return True

    @staticmethod
    def uniq_chars(string, method):
        """Selects algorithm to determine redundancy of chars in given string

        Keyword arguments:
        string -- str to check for redundancy of chars
        method -- algorithm to use on string
        """
        if not isinstance(method, int):
            raise AttributeError(
                "Given input, {}, must be an instance of int".format(method))
        selected = UniqChars.__METHODS.get(method)
        if not selected:
            raise AttributeError("Invalid method input: '{}'".format(method))
        return UniqChars.__METHODS[method](UniqChars(), string)

    SORTED = 1
    BOOL = 2
    BIT = 3

    __METHODS = {
        1: __method1,
        2: __method2,
        3: __method3
    }
